_posts/2016-03-12-make-your-data-obey.markdown

```diff
@@ -0,0 +1,187 @@
+obey.creator('uuidCreator', () => uuid.v4())
+
+const user = obey.model({
+    // The id now specifies that the creator 'uuid' should be used if no value is present
+    id: { type: 'uuid', creator: 'uuidCreator' required: true },
+    email: { type: 'email', required: true },
+    password: { type: 'string', require: true }
+})
+{% endhighlight %}
+
+So, now we just pass in the `email` and `password` and our `id` is auto-generated.
+
+But that `password` is still not cool; we don't want to save that in plain-text...
+
+## Modifiers
+
+Obey wasn't just built for validation, but also to support coercion and modification. Modifiers allow for doing this in a clean, simple method. Let's revisit the model again:
+
+{% highlight javascript %}
+import obey from 'obey'
+import uuid from 'node-uuid'
+// Import the crypto module
+import crypto from 'crypto'
+
+obey.creator('uuid', () => uuid.v4())
+
+// Add a modifier that returns md5 encrypted value
+obey.modifier('encrypt', val => crypto.createHash('md5').update(val).digest('hex'))
+
+const user = obey.model({
+    id: { type: 'uuid', creator: 'uuidCreator' required: true },
+    email: { type: 'email', required: true },
+    // Password now specifies the modifier 'encrypt' should be used on the value
+    password: { type: 'string', modifier: 'encrypt', require: true }
+})
+{% endhighlight %}
+
+Alright, so now not only does our validation create a UUID, it also encrypts that password for us. We've changed the model to accomplish this, but we validate exactly the same, Obey handles this all automatically based on the modelling we defined in one location.
+
+BUT, what about that `email`? Probably want to make sure it's unique...
+
+## Types
+
+Having that `email` contain one that's already in the datasource would be bad news bears right? So here's the point where a lot of developers say "my database will reject it!". Good point, not so good execution though.
+
+You want your datasource to go into (a potentially locking) `insert` state to do your validation for you? Really? Why not do that all inside the validation? It is validating data after all. Luckily `types` are customizable, so let's make our own:
+
+{% highlight javascript %}
+import obey from 'obey'
+import uuid from 'node-uuid'
+import crypto from 'crypto'
+// Here's our db reference again...
+import db from './db'
+
+obey.creator('uuid', () => uuid.v4())
+
+obey.modifier('encrypt', val => crypto.createHash('md5').update(val).digest('hex'))
+
+// Add a type to ensure we have a valid, unique email
+obey.type('uniqueEmail', context => {
+    // Read from the 'db' datasource where email is the value
+    return db.read({ email: context.val }).then(res => {
+        // Fail if it already exists
+        if (res.length !== 0) context.fail('Email must be unique')
+        // Fail if it's not a legit email
+        const emailRegEx = /^([a-zA-Z0-9_\.\-])+\@(([a-zA-Z0-9\-])+\.)+([a-zA-Z0-9]{2,4})+$/
```

Comment:
While I agree with your point, the point of the post is to show a _clear_ example of Obey. The goal of the example is to demonstrate the async call to the db, I feel like putting in a more complex validation check just distracts from what I'm trying to convey.